### PR TITLE
Worksheet Connections

### PR DESCRIPTION
**Pull request by darth-cheney**
## What ##
This PR implements a new custom element called `<ws-connection>` that specifies relationships between Worksheets. Additionally, it uses the [leader-line](https://github.com/anseki/leader-line) library to draw graphical connections between Worksheets using dynamic SVGs.
  
## Implementation ##
Proper use of a `<ws-connection>` element is simple: specify both a `source=` and `destination=` value representing the `id` of the respective Worksheet elements, and a connection will be drawn between them. This visual representation of the connection will update as the Worksheet objects are moved around the workspace.
  
Part of the advantage of this design is that we can use normal DOM queries to figure out the connections between different Worksheets.
  
### `WSPort` ###
Some work went into the idea of "ports" which would allow users to dynamically great connections between Worksheets by dragging the mouse from the port of one sheet to the port of a target sheet. It became unclear whether or not this would even be useful to users, or what creating connections in this way would even achieve compared to other methods we have planned in the near future. I have left that code in this PR for posterity, in case we ever come back to it. I suggest that after some reasonable amount of time, if we have note revisited this concept, we remove it from the repository.

### CODE REVIEW

**Comment by dkrasner** [comment]
Cool; a few questions:

is the leader line supposed to stick in `ws-port-basic.html`. i keep dropping it over the port and nothing happens, ie it disappears after a second or so?

was there supposed to be a way to do this with worksheets (not just in the example)? I recall you added a port-icon, but don't see it in the UI. (The diff on Worksheet.js got messed up somehow in this PR so it's hard to see what was changed). 

do we want to reconcile this with the current linker? right now in theory we have two concurrent ways to "link" elements

File: js/WSConnection.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+    connectedCallback(){
+        if(this.isConnected){
+            this.updateLinkedSheet(
+                "",
+                this.getAttribute("source")
+            );
+            this.updateLinkedSheet(
+                "",
+                this.getAttribute("destination")
```
Comment (dkrasner) [inline js/WSConnection.js:26]: up until now we call this attribute "target" (source -> target). It's not so important what we call but there is already some code out there - css class toggles, and how the instructions are processed - which depends on it.

**Comment by darth-cheney** [comment]
> is the leader line supposed to stick in ws-port-basic.html. i keep dropping it over the port and nothing happens, ie it disappears after a second or so?
  
This works for me. Are you dropping on the other port (and not the container for the other port)?
  
> was there supposed to be a way to do this with worksheets (not just in the example)? I recall you added a port-icon, but don't see it in the UI. (The diff on Worksheet.js got messed up somehow in this PR so it's hard to see what was changed).
  
I removed this because we determined that we would not have users connect worksheets this way. We can always bring it back later if we want.
  
> do we want to reconcile this with the current linker? right now in theory we have two concurrent ways to "link" elements
  
Let's do this in a subsequent pairing session, if you're up for it.

File: js/WSConnection.js (hunk 1 above)
Comment (darth-cheney) [reply js/WSConnection.js]: Fixed in https://github.com/dkrasner/worksheet/pull/23/commits/e41dd7a47f28cd27e3d7ac133295ebd11b6f3ed1

**Comment by dkrasner** [comment]
> > is the leader line supposed to stick in ws-port-basic.html. i keep dropping it over the port and nothing happens, ie it disappears after a second or so?
> 
> This works for me. Are you dropping on the other port (and not the container for the other port)?

this is what i am doing
image
 
and nothing happens, ie the arrow just disappears

(also i've tried dropping everywhere) i am on FF

**Comment by darth-cheney** [comment]
This is what I'm seeing in FF:
https://user-images.githubusercontent.com/1640299/184405804-04ec4eb8-93ba-49fb-91f8-f3bd7dc29cf8.mov

**Comment by dkrasner** [comment]
yeah well something is wrong i can't get that to happen - no errors nothing

**Comment by dkrasner** [comment]
https://user-images.githubusercontent.com/1154326/184627245-d69b091f-9165-413a-819c-9046741f09a2.mov

**Comment by dkrasner** [comment]
merging with open issue #26 related to above

**Pull request merged**